### PR TITLE
Add better simplifier tracing and fix unsound simplification step

### LINKED ISSUE #141: Debug simplifier failures
The deep formula simplification pass has been observed to (rarely) generate simplified terms not equivalent to the originals. It has been instrumented to report when this happens (via Debug.Trace). We need to figure out which simplification rule is at fault.

### PR DESCRIPTION
**Pull request by danmatichuk**
fixes #141

### CODE REVIEW

File: src/What4/ExprHelpers.hs [hunk 1]
```diff
@@ -973,36 +1005,72 @@ simplifyBVOpInner sym go app = case app of
...
+    f app = case simplifyBVOpInner sym (\e1 e2 -> inIO (simp_check e1 e2)) (\e' -> inIO (go e')) app of
+      Just g -> Just <$> liftIO g
+      Nothing -> return Nothing
+
+    go :: forall tp'. W4.SymExpr sym tp' -> m (W4.SymExpr sym tp')
+    go e = W4B.idxCacheEval cache e $ simplifyApp sym simp_check f e
+
+  inIO (go outer)
+
+simplifyApp ::
```
Comment (travitch) [inline src/What4/ExprHelpers.hs:1038]: What are the expected uses/semantics of the "double-check simplification step" - is this a full simplification validation?

File: src/Pate/Verification/Simplify.hs [hunk 2]
```diff
@@ -54,6 +57,47 @@ simplifySubPreds a = withValid $ withSym $ \sym -> do
...
+  W4.SymExpr sym tp ->
+  EquivM sym arch (W4.SymExpr sym tp)
+simplifyBVOps_trace sym checkWork outer = do
+  cache <- W4B.newIdxCache
+  let
+    go :: forall tp'. W4.SymExpr sym tp' -> EquivM_ sym arch (W4.SymExpr sym tp')
+    go e = W4B.idxCacheEval cache e $ WEH.simplifyBVOps' sym checkWork e
+  go outer
+
+simplifyWithSolver ::
```
Comment (travitch) [inline src/Pate/Verification/Simplify.hs:78]: What does this use the solver for? What is the behavior if there is a timeout?

File: src/Pate/Verification/Simplify.hs [hunk 3]
```diff
@@ -3,10 +3,13 @@
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
```
Comment (travitch) [inline src/Pate/Verification/Simplify.hs:6]: Is this extension actually used? I didn't see any type applications at a glance

File: src/What4/ExprHelpers.hs (hunk 1 above)
Comment (danmatichuk) [reply src/What4/ExprHelpers.hs:1060]: I wrapped up this check in a newtype and added some comments

File: src/Pate/Verification/Simplify.hs (hunk 2 above)
Comment (danmatichuk) [reply src/Pate/Verification/Simplify.hs:84]: Added a comment explaining this

File: src/Pate/Verification/Simplify.hs (hunk 3 above)
Comment (danmatichuk) [reply src/Pate/Verification/Simplify.hs]: Removed

**Review by travitch — APPROVED** [review]
Thanks, looks good

**Pull request merged**